### ISSUE TITLE
Model does not converge after saving and loading

### ISSUE THREAD

**Issue by yangw1234** [keras-team/keras#16023]
**System information**.
- Have I written custom code (as opposed to using a stock example script provided in Keras): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Linux
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): 2.8.0
- Python version: 3.7.10
- Bazel version (if compiling from source):
- GPU model and memory: cpu
- Exact command to reproduce:

**Describe the problem**.

Keras model does not converge after saving and loading.

**Describe the current behavior**.
After calling `model.save(...)` and `model = tf.keras.models.load_model(...)`, the model failed to converge.

**Describe the expected behavior**.
Adding `model.save(...)` and `model = tf.keras.models.load_model(...)` should not effect the training process.

- Do you want to contribute a PR? (yes/no):  No

**Standalone code to reproduce the issue**.

```python
import tensorflow as tf
mnist = tf.keras.datasets.mnist

(x_train, y_train),(x_test, y_test) = mnist.load_data()
x_train, x_test = x_train / 255.0, x_test / 255.0

model = tf.keras.models.Sequential([
  tf.keras.layers.Flatten(input_shape=(28, 28)),
  tf.keras.layers.Dense(128, activation='relu'),
  tf.keras.layers.Dropout(0.2),
  tf.keras.layers.Dense(10, activation='softmax')
])

model.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])
########### added lines ###########
model.save("/tmp/mnist_model")
model = tf.keras.models.load_model('/tmp/mnist_model')
################################

model.fit(x_train, y_train, epochs=5)
model.evaluate(x_test, y_test)
```
**Source code / logs**.

After adding saving and loading:
```
2022-02-08 09:03:19.403489: I tensorflow/core/platform/cpu_feature_guard.cc:151] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 AVX512F FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
2022-02-08 09:03:19.698194: W tensorflow/python/util/util.cc:368] Sets are not currently considered sequences, but this may change in the future, so consider avoiding using them.
Epoch 1/5
1875/1875 [==============================] - 6s 3ms/step - loss: 0.2983 - accuracy: 0.1004
Epoch 2/5
1875/1875 [==============================] - 6s 3ms/step - loss: 0.1451 - accuracy: 0.0992
Epoch 3/5
1875/1875 [==============================] - 6s 3ms/step - loss: 0.1069 - accuracy: 0.0990
Epoch 4/5
1875/1875 [==============================] - 6s 3ms/step - loss: 0.0881 - accuracy: 0.0991
Epoch 5/5
1875/1875 [==============================] - 6s 3ms/step - loss: 0.0754 - accuracy: 0.0989
313/313 [==============================] - 1s 2ms/step - loss: 0.0775 - accuracy: 0.0992
```

Remove saving and loading:

```
2022-02-08 09:05:18.683461: I tensorflow/core/platform/cpu_feature_guard.cc:151] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 AVX512F FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
Epoch 1/5
1875/1875 [==============================] - 6s 3ms/step - loss: 0.3010 - accuracy: 0.9133   
Epoch 2/5
1875/1875 [==============================] - 6s 3ms/step - loss: 0.1452 - accuracy: 0.9574
Epoch 3/5
1875/1875 [==============================] - 6s 3ms/step - loss: 0.1099 - accuracy: 0.9666
Epoch 4/5
1875/1875 [==============================] - 6s 3ms/step - loss: 0.0878 - accuracy: 0.9729
Epoch 5/5
1875/1875 [==============================] - 6s 3ms/step - loss: 0.0746 - accuracy: 0.9766
313/313 [==============================] - 1s 2ms/step - loss: 0.0744 - accuracy: 0.9770
```

**Comment by yangw1234** [comment]
It seems we cannot call `model.save()` before calling `model.fit`.

**Comment by chunduriv** [comment]
Here `model.fit` trains the model for a fixed number of epochs  on a dataset. We should use `model.save` to save the model to Tensorflow `SavedModel` or a single `HDF5` file after training. Thanks!

**Comment by yangw1234** [comment]
> Here `model.fit` trains the model for a fixed number of epochs on a dataset. We should use `model.save` to save the model to Tensorflow `SavedModel` or a single `HDF5` file after training. Thanks!

Hi @chunduriv , I understand how `mode.fit` and `model.save` works. **I am here to report a bug.**

If I understand correctly, adding the following two lines will not affect the model's state.
```
model.save("/tmp/mnist_model")
model = tf.keras.models.load_model('/tmp/mnist_model')
```

However, the current behavior is adding the above two lines will cause a model not to converge.

You can easily reproduce the problem by running the code I posted.

**Comment by chunduriv** [comment]
@yangw1234,

Generally [model.save](https://www.tensorflow.org/api_docs/python/tf/keras/Model#save) will save a model's architecture, weights and training configuration in a single file/folder. 

Here to get correct values after loading the model, can you configure the model for training before `model.fit` as shown below

```
model = tf.keras.models.load_model('/tmp/mnist_model') # load model
model.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])
model.fit(x_train, y_train, epochs=5)
model.evaluate(x_test, y_test)
```

Please refer attached [gist](https://colab.research.google.com/gist/chunduriv/9b9f748c2e4ceb7f9ada226b50ed5e09/16023.ipynb) for reference. For more details please refer [this](https://www.tensorflow.org/tutorials/keras/save_and_load) tutorial. Thanks!

**Comment by yangw1234** [comment]
> 

Thanks @chunduriv , I see what you are saying, but I have one more questions. 

Will the `model.compile` line in your example overwrite the optimizer states saved by `model.save`?

**Comment by yangw1234** [comment]
@chunduriv actually, it seems calling `model.save` itself will **also** make the current model not converge.

![image](https://user-images.githubusercontent.com/10561966/154002790-20c4249d-ef90-41e8-81bd-8c1be8bca5e8.png)

**Comment by yangw1234** [comment]
I opened a pr to fix this issue #16069

**Comment by yangw1234** [comment]
I think the root cause is here:
https://github.com/keras-team/keras/blob/master/keras/saving/saving_utils.py#L326.

Here it uses `model.outputs` as the target to build metrics. In the build logic, however, the target shape is used to determine the right kind of accuracy (see [here](https://github.com/keras-team/keras/blob/master/keras/engine/compile_utils.py#L546-L569)). As a result, the wrong accuracy is built into the model. So the model training is fine, but the accuracy calculation is wrong.
